### PR TITLE
Run without sudo if not needed

### LINKED ISSUE #75: sudo: command not found
Many docker containers don't come with sudo. I'm also using [UTM](https://mac.getutm.app/) to run Linux and their ArchLinux image is missing sudo (it's running as root!). I can install sudo, but I thought I'd mention it. Maybe setup could check for sudo before trying to run it.

### PR DESCRIPTION
**Pull request by TanmayPatil105**
## What is the PR about
Solved `sudo: command not found`

## Issue Solved
Fix #75

### CODE REVIEW

File: install.sh
Comment (mxcl) [inline install.sh:273]: it this enough? Can the user be named otherwise? IIRC user-id 0 is always “root” can we check that?

File: install.sh
Comment (mxcl) [inline install.sh:275]: well if the user is `root` presumably we should just not use `sudo` since the `if` checks for either and `sudo` may not be there in fact.

File: install.sh
Comment (TanmayPatil105) [reply install.sh:273]: Yess, It can be resolved by using `$EUID` instead of `$USER` and changing the condition to:
```
if [ "$EUID" -eq 0 ]
```

Comment (jhheider) [reply install.sh:273]: Technically, we don't care if they're root, but only if they have write access to the directory (or one of its parents, if we have to mkdir). In practice this means root, but it might not if I `sudo chown -R jacob /usr/local`.

Comment (TanmayPatil105) [reply install.sh:273]: We only need the write access and the user need not be a root. So, the changes would be ..
```
if [ -w /usr/local ]
```
So, the code would finally look like
```
if [ -w /usr/local ];
then
	mkdir -p /usr/local/bin
	ln -sf "$tea" /usr/local/bin/tea
else
	if which sudo >/dev/null;
	then
		sudo mkdir -p /usr/local/bin
		sudo ln -sf "$tea" /usr/local/bin/tea
	else
		echo  #spacer
		gum format -- <<-EOMD
			> hmmm, sudo command not found.
                        > try installing sudo
			EOMD
	fi
fi
```

Comment (jhheider) [reply install.sh:273]: This should work fine. I think to deal with all the use cases, the logic might be something like this:

if write(/u/l/b) or (write(/u/l) and not(/u/l/b)) or (write (/u) and not(/u/l)) or (write(/) and not(/u)) then mkdir and ln
elif sudo then mkdir and ln
else report an issue

Comment (TanmayPatil105) [reply install.sh:273]: I think there should be only a `write(/u/l/b)` because suppose the parent directory is owned by `user`, the child directory  may or may not be owned by root. So, `write(/u/l/b/)` is must.

File: install.sh
Comment (TanmayPatil105) [reply install.sh:279]: Fixed in recent commit!

File: install.sh
Comment (jhheider) [reply install.sh]: but if it doesn't exist, then writing /usr/local is sufficient. and if that doesn't exist, writing /usr is sufficient. etc. that's the issue.

Comment (jhheider) [reply install.sh]: perhaps the logical is is `if -w /usr/local/bin || mkdir -p /usr/local/bin`

Comment (TanmayPatil105) [reply install.sh]: Makes sense! Addressing it shortly.

Comment (jhheider) [reply install.sh]: We should add a stanza to `ci.yml` that tests something like:
```sh
sudo su - root
dpkg -P sudo
$GITHUB_WORKSPACE/install.sh
```
to ensure it functions like we think.

Comment (TanmayPatil105) [reply install.sh]: Should the job be like:
```
root:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Run as root
        run: sudo su - root && dpkg -P sudo
      - name: Installation
        run: $GITHUB_WORKSPACE/install.sh
```

Comment (TanmayPatil105) [reply install.sh]: Done!

File: install.sh
Comment (jhheider) [inline install.sh:273]: I'm not sure this is precisely correct, but you'd have to do something very esoteric to break it.
```sh
sudo mkdir -p /usr/local/bin
chmod 555 /usr/local/bin
chmod 777 /usr
```

Comment (TanmayPatil105) [reply install.sh:273]: Works!

Comment (TanmayPatil105) [reply install.sh:273]: Works!

**Comment by jhheider** [comment]
Looks like the check might need to be more like...
```sh
sudo su - root -c "dpkg -P sudo && $GITHUB_WORKSPACE/install.sh"
```

File: install.sh
Comment (jhheider) [inline install.sh:281]: I wasn't recommending we add this (it doesn't make much sense); I was suggesting if you had done something like this it would pass the writability test, even though it's not writable.

Comment (TanmayPatil105) [reply install.sh:281]: Since we are using `sudo`, writability test won't be a problem or will it be.
It would be better if I remove it.

**Comment by TanmayPatil105** [comment]
> https://github.com/teaxyz/setup/actions/runs/3518451655/jobs/5897390969
> 
> Might need `--force` or `--force-all`. At least we can "break" the VM that runs the check, since they're destroyed after.

Should it be like:
```
run: sudo su - root -c "dpkg -P sudo --force-all && $GITHUB_WORKSPACE/install.sh"
```

File: install.sh
Comment (jhheider) [reply install.sh:281]: This should not be present. I was saying that *if* /usr/local/bin exists and is _not_ writable, but /usr _is_ writable, it will pass your test, but otherwise fail. I think the correct test is:
```sh
if [ -w /usr/loca/bin ] || mkdir -p /usr/local/bin 2>&1; then
```

**Comment by jhheider** [comment]
> Should it be like:
> 
> ```
> run: sudo su - root -c "dpkg -P sudo --force-all && $GITHUB_WORKSPACE/install.sh"
> ```

That will probably work (assuming `dpkg`) isn't overly positional. The [manpage](https://man7.org/linux/man-pages/man1/dpkg.1.html) suggests `dpkg --force-all -P sudo` _should_ definitely work.

File: install.sh
Comment (TanmayPatil105) [reply install.sh:281]: Like this:
```
if [ -w /usr/local/bin ] || mkdir -p /usr/local/bin ;
then
	ln -sf "$tea" /usr/local/bin/tea
elif which sudo >/dev/null;
then
	sudo ln -sf "$tea" /usr/local/bin/tea
else
	echo  #spacer
	gum format -- <<-EOMD
		> hmmm, sudo command not found.
		> try installing sudo
		EOMD
fi
```
right?

Comment (jhheider) [reply install.sh:281]: I think that is likely correct. In English, that's "if we can write to /usr/local/bin **OR** we can create /usr/local/bin", which I think are the two sufficient conditions.  Note that you'll also need the mkdir in the `elif` block, since it will have failed without `sudo` in the `if` statement. It's also best practice to `2>&1 >/dev/null` in `if` tests to prevent unnecessary text leaking to output.

Comment (TanmayPatil105) [reply install.sh]: I tried runnig a test workflow to change user to sudo and the user renamed the same.
```
Run sudo su - root -c "echo '$EUID'"
1001
```

Comment (TanmayPatil105) [reply install.sh]: I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?

Comment (jhheider) [reply install.sh]: > I tried runnig a test workflow to change user to sudo and the user renamed the same.
> ```
> Run sudo su - root -c "echo '$EUID'"
> 1001
> ```

I think you need `'echo $EUID'` but `sudo su - root -c id` might be more sure.

Comment (jhheider) [reply install.sh]: > I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?

Maybe, but removing sudo ensures it's unavailable.

Comment (TanmayPatil105) [reply install.sh]: It worked.
```
Run sudo su - root -c id
uid=0(root) gid=0(root) groups=0(root)
```

Comment (TanmayPatil105) [reply install.sh]: > > I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?
> 
> Maybe, but removing sudo ensures it's unavailable.

Removing `sudo` might be the only solution to solve workflow error ig.

Comment (jhheider) [reply install.sh]: `sudo su -` should work, but maybe try a debian image instead if not.

Comment (TanmayPatil105) [reply install.sh]: `sudo su-` worked but the it shows the following error now
```
sudo su -c 'ls && $GITHUB_WORKSPACE/install.sh'
```
```
/home/runner/work/setup/setup
LICENSE.txt
README.md
action.js
action.yml
install.sh
bash: /install.sh: No such file or directory
```

Comment (TanmayPatil105) [reply install.sh]: Ahh! missed it. 
Tried it and now the issue if of executing install.sh
```
sudo su -c './install.sh'
```
```
./install.sh: 124: shift: can't shift that many
```

Comment (jhheider) [reply install.sh]: This is coming up in other contexts, such as [this](https://github.com/teaxyz/cli/issues/178), so we may have a regression.

Comment (TanmayPatil105) [reply install.sh]: Yupp, `shift` error generally comes up when we use `case` `esac`.

Comment (jhheider) [reply install.sh]: I can take a look as well, but running `sudo su -c 'bash -x install.sh'` should help see where we're breaking down.

Comment (TanmayPatil105) [reply install.sh]: It went into an infinite loop.
```
+ set -e
+ set -o noglob
+ prepare
+ which tar
+ test -n ''
+ test 0 -eq 0
+ MODE=install
++ uname
++ uname -m
+ HW_TARGET=Linux/x86_64
+ ZZ=gz
+ case $HW_TARGET in
+ MIDFIX=linux/x86-64
++ uname
+ test Linux = Darwin
+ test gz = gz
+ which base64
+ BASE64_TARXZ=/Td6WFoAAATm1rRGAgAhARYAAAB0L+Wj4AX/AFNdADMb7AG6cMNAaNMVK8FvZMaza8QKKTQY6wZ3kG/F814lHE9ruhkFO[5](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:6)DAG7XNamN7JMHavgmbbLacr72NaAzgGUXOstqUaGb[6](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:7)kbp[7](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:8)jrkF+3aQT12CAAB[8](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:9)Uikc1gG8RwABb4AMAAAAeGbHwbHEZ/sCAAAAAARZWg==
+ echo /Td6WFoAAATm1rRGAgAhARYAAAB0L+Wj4AX/AFNdADMb7AG6cMNAaNMVK8FvZMaza8QKKTQY6wZ3kG/F814lHE[9](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:10)ruhkFO5DAG7XNamN7JMHavgmbbLacr72NaAzgGUXOstqUaGb6kbp7jrkF+3aQT[12](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:13)CAAB8Uikc1gG8RwABb4AMAAAAeGbHwbHEZ/sCAAAAAARZWg==
+ base64 -d
+ tar Jtf -
+ ZZ=xz
+ case "$ZZ" in
+ TAR_FLAGS=xJ
+ test -n '' -a -f '/tea.xyz/v*/bin/tea'
+ test -z ''
+ which tea
+ test -z ''
+ test install = exec
+ TEA_PREFIX=/root/.tea
+ test -z ''
+ which curl
+ CURL='curl -Ssf'
+ get_gum
+ test '!' -t 1 -o '' = 0

+ GUM=gum_no_tty
+ test install = install -a -z ''
+ welcome
+ gum format --
+ case "$1" in
+ gum_no_tty format --
+ cmd=format
+ test format '!=' --
+ shift
+ test -- '!=' --
```
And the workflow is not terminating too :/

Comment (jhheider) [reply install.sh]: yes, I think possibly gum -- spin is looping when there's no TTY.

Comment (TanmayPatil105) [reply install.sh]: Possible solutions?

Comment (jhheider) [reply install.sh]: Looking into it now. I assume you're running via SSH (or other non-tty method)?

Comment (TanmayPatil105) [reply install.sh]: Not sure about it 😅!

Comment (TanmayPatil105) [reply install.sh]: Try
```
test $(( $# )) && shift
```
Not sure if it will work.

Comment (jhheider) [reply install.sh]: I'm hopeful that https://github.com/teaxyz/setup/pull/82 will fix the loop.

Comment (jhheider) [reply install.sh]: Since it's a no-tty issue (running under sudo su), you might want to merge in _that_ branch and see if that resolves everything here.

Comment (TanmayPatil105) [reply install.sh]: Yupp! Working fine.

Comment (TanmayPatil105) [reply install.sh]: Not sure why it is falling again :/. It worked fine on my branch.

Comment (jhheider) [reply install.sh]: It doesn't look like this branch has merged in `teaxyz:fix-infinite-loop-no-tty`.

Comment (TanmayPatil105) [reply install.sh]: Ahh! It is not merged that's why.

Comment (TanmayPatil105) [reply install.sh]: Ready to merge I guess!

Comment (TanmayPatil105) [reply install.sh]: We need to install gum in VMs in order to pass the workflows.
So, for macOS
```
brew install gum
```
```
apt install gum
```

Comment (jhheider) [reply install.sh]: I think if you merge the newest commit to main, it should work.

File: install.sh
Comment (jhheider) [inline install.sh:282]: This line needs to be `gum_func` as of ff81688b1398760b3ea0ea8f03f4bc9193a028a2. Otherwise, I think we've gotten where we want to! Party!

Comment (TanmayPatil105) [reply install.sh]: Done!

**Review by mxcl — APPROVED** [review]

**Comment by mxcl** [comment]
@jhheider please squash once approved

File: install.sh
Comment (TanmayPatil105) [reply install.sh]: Resolved conflicts!!
Following job was added to ci.yml
```
specific-version:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: ./
        with:
          version: 0.13.8
      - run:
          test "$(tea --version)" = "tea 0.13.8"
```

**Review by jhheider — APPROVED** [review]
Looks great!

File: .github/workflows/ci.yml
Comment (mxcl) [inline .github/workflows/ci.yml:143]: does this test anything? CI runs as root by default.

Shouldn’t we instead create a new user and then test it runs with `sudo` and then change the permissions so that user can write to `/usr/local/bin` and then uninstall sudo and check it works?

**Comment by mxcl** [comment]
so yeah, the CI addition didn't test anything and actually this PR broke the installer. Errors with:

```
ln: /usr/local/bin/tea: Permission denied
```

**Comment by mxcl** [comment]
yeah it's because `mkdir -p` returns `0` if the directory already exists

**Comment by mxcl** [comment]
Fixed in #91

**Comment by jhheider** [comment]
Well, shoot. And here I thought I was being clever. I don't think your fix is 100%, however, since if `/usr/local` doesn't exist, but we can write `/usr`, but `sudo` isn't installed, we'll get a failure (right? I think I did that correctly.)

It probably needs to be `test -w /usr/local/bin -o ((-e /usr/local -a -w /usr/local) -o (-e /usr -a -w /usr) -o -w /)` which is just awful. So, maybe, `test -w /usr/local/bin -o ! -e /usr/local/bin -a mkdir -p /usr/local/bin >/dev/null 2>&1`?

**Comment by mxcl** [comment]
Yep you're right, pls PR. If you can be arsed. This is pretty edge case I reckon (not having a `/usr/local` dir that is).

**Comment by mxcl** [comment]
side note, `/usr/local` who decided on `local` it's so not 3 characters.

**Comment by jhheider** [comment]
It is an extremely edge case. But someone will install us in some microcontroller with a deeply-custom linux install and we'll bork.

**Comment by mxcl** [comment]
I am deeply happy about your commitment to “someone will experience this” and you're 100% right.

**Comment by jhheider** [comment]
https://github.com/teaxyz/setup/pull/95

**Pull request merged**
